server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractPercentilesAggregationBuilder.java

```diff
@@ -0,0 +1,361 @@
+ * It provides a set of common fields/functionality for setting the available algorithms (TDigest and HDRHistogram),
+ * as well as algorithm-specific settings via a {@link PercentilesConfig} object
+ */
+public abstract class AbstractPercentilesAggregationBuilder<T extends AbstractPercentilesAggregationBuilder<T>>
+    extends ValuesSourceAggregationBuilder.LeafOnly<ValuesSource, T> {
+
+    public static final ParseField KEYED_FIELD = new ParseField("keyed");
+    protected boolean keyed = true;
+    protected double[] values;
+    private PercentilesConfig percentilesConfig;
+    private ParseField valuesField;
+
+    public static <T extends AbstractPercentilesAggregationBuilder<T>> ConstructingObjectParser<T, String> getParser(String aggName,
+                                                         TriFunction<String, double[], PercentilesConfig, T> fn,
+                                                         ParseField valuesField) {
+
+        /**
+         * This is a non-ideal ConstructingObjectParser, because it is a compromise between Percentiles and Ranks.
+         * Ranks requires an array of values because there is no sane default, and we want to keep that in the ctor.
+         * Percentiles has defaults, which means the API allows the user to either use the default or configure
+         * their own.
+         *
+         * The mutability of Percentiles keeps us from having a strict ConstructingObjectParser, while the ctor
+         * of Ranks keeps us from using a regular ObjectParser.
+         *
+         * This is a compromise, in that it is a ConstructingOP which accepts all optional arguments, and then we sort
+         * out the behavior from there
+         *
+         * `args` are provided from the ConstructingObjectParser in-order they are defined in the parser.  So:
+         *  - args[0]: values
+         *  - args[1]: tdigest config options
+         *  - args[2]: hdr config options
+         *
+         *  If `args` is null or empty, it means all were omitted.  This is usually an anti-pattern for
+         *  ConstructingObjectParser, but we're allowing it because of the above-mentioned reasons
+         */
+        ConstructingObjectParser<T, String> parser = new ConstructingObjectParser<>(aggName, false, (args, name) -> {
+
+            if (args == null || args.length == 0) {
+                // Note: if this is a Percentiles agg, the null `values` will be converted into a default,
+                // whereas a Ranks agg will throw an exception due to missing a required param
+                return fn.apply(name, null, new PercentilesConfig.TDigest());
+            }
+
+            double[] values = args[0] != null ? ((List<Double>) args[0]).stream().mapToDouble(Double::doubleValue).toArray() : null;
+            PercentilesConfig percentilesConfig;
+
+            if (args[1] != null && args[2] != null) {
```

Comment:
I think it'd be a little more clear if you did:
```
PercentilesConfig explicitTDigestConfig = (PercentilesConfig) args[1];
PercentilesConfig explicitHdrConfig = (PercentilesConfig) args[2];
if (explicitTDigestConfig == null && explicitHdrConfig == null) {
...
```

IT feels like it'd be cleaner to throw an exception if both are set. Is that something you aren't doing for bwc? Maybe leave a comment in that case.

Reply:
++ I can add some local variables to make it more explicit.


> IT feels like it'd be cleaner to throw an exception if both are set. Is that something you aren't doing for bwc? Maybe leave a comment in that case.

That's exactly what this block is doing :)  Probably will be clearer after the explicit local variables
```
if (args[1] != null && args[2] != null) {
  throw new IllegalArgumentException("Only one percentiles method should be declared.");
}
```